### PR TITLE
Goshawk tweaks

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: gossip/minimal/goshawk.go [hunk 1]
```diff
@@ -173,30 +156,28 @@ func (hawk *Goshawk) Fly(ctx context.Context) {
 			glog.Infof("finished Checker(%s)", origin.Name)
 		}(origin)
 	}
+
 	wg.Wait()
 }
 
 // Scanner runs a continuous scan of the destination log.
 func (hawk *Goshawk) Scanner(ctx context.Context) {
```
Comment (pavelkalinnikov) [inline gossip/minimal/goshawk.go]: nit: Could the method name be a verb, e.g. `Scan`?

File: gossip/minimal/goshawk.go [hunk 2]
```diff
@@ -173,30 +156,28 @@ func (hawk *Goshawk) Fly(ctx context.Context) {
...
-		hawk.scanOpts.StartIndex = size
-
-		glog.V(2).Infof("Scanner(%s): wait for a %s tick", hawk.dest.Name, hawk.dest.MinInterval)
-		select {
-		case <-ctx.Done():
-			glog.Infof("Scanner(%s): termination requested", hawk.dest.Name)
-			return
-		case <-ticker.C:
+	fetcher := scanner.NewFetcher(hawk.dest.Log, &hawk.fetchOpts)
+	fetcher.Run(ctx, func(batch scanner.EntryBatch) {
```
Comment (pavelkalinnikov) [inline gossip/minimal/goshawk.go]: `Run` returns an error. Maybe return this error from `Scanner()`?

File: gossip/minimal/goshawk.go [hunk 3]
```diff
@@ -173,30 +156,28 @@ func (hawk *Goshawk) Fly(ctx context.Context) {
...
+		glog.V(2).Infof("Scanner(%s): examine batch [%d, %d)", hawk.dest.Name, batch.Start, int(batch.Start)+len(batch.Entries))
+		for i, entry := range batch.Entries {
+			index := batch.Start + int64(i)
+			rawLogEntry, err := ct.RawLogEntryFromLeaf(index, &entry)
+			if rawLogEntry == nil {
```
Comment (pavelkalinnikov) [inline gossip/minimal/goshawk.go]: Are you sure you don't want to check `err != nil` here? Maybe check both?

File: gossip/minimal/goshawk.go [hunk 4]
```diff
@@ -74,27 +74,23 @@ func NewGoshawkFromFile(ctx context.Context, filename string, hc *http.Client, s
...
 	if cfg.SourceLog == nil || len(cfg.SourceLog) == 0 {
 		return nil, errors.New("no source log config found")
 	}
-	bufSize := cfg.BufferSize
-	if bufSize <= 0 {
-		// Allow at least one buffered STH in the channel so context cancellation
-		// works (everything shuts down at the same time, so the fetcher might otherwise
-		// block on a channel that no longer has anything reading from it).
-		bufSize = 1
+	if cfg.BufferSize < 0 {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply gossip/minimal/goshawk.go]: I don't see the point of introducing a new local variable that's got almost the same name as the existing field.

File: gossip/minimal/goshawk.go (hunk 1 above)
Comment (daviddrysdale) [reply gossip/minimal/goshawk.go]: No, the name is deliberate and consistent -- all of the methods that do something continuously or continually (and so get run in goroutines) have names that are nouns built from a verb-subject -- `Checker`, `Retriever`, `Scanner`, `Submitter`.

File: gossip/minimal/goshawk.go (hunk 3 above)
Comment (daviddrysdale) [reply gossip/minimal/goshawk.go]: I don't see the point -- `RawLogEntryFromLeaf` complies with normal Go conventions and always returns a `nil` entry if `err` is non-nil, and the key check for the code below is whether there's a `RawLogEntry` to play with.

File: gossip/minimal/goshawk.go (hunk 2 above)
Comment (daviddrysdale) [reply gossip/minimal/goshawk.go]: Good idea - if the fetcher fails (for something other than context cancellation) the other goroutines will carry on regardless.

**Comment by daviddrysdale** [comment]
PTAL; I've added another commit -- on testing more closely, it looks like https://github.com/google/trillian/pull/1333 has broken continuous mode, because a normal `error` changed from being retryable to *not* being retryable.  Will this affect Migrillian too?

File: scanner/fetcher.go [hunk 5]
```diff
@@ -230,7 +231,8 @@ func (f *Fetcher) updateSTH(ctx context.Context) error {
 
 		quick := time.Now().Before(quickDeadline)
 		if sth.TreeSize <= lastSize || quick && sth.TreeSize < targetSize {
-			return errors.New("waiting for bigger STH")
+			// Use an explicitly retriable error code.
+			return status.Errorf(codes.Unavailable, "wait for bigger STH than %d (last=%d, target=%d)", sth.TreeSize, lastSize, targetSize)
```
Comment (gdbelvin) [inline scanner/fetcher.go]: Thanks for making this change

File: gossip/minimal/goshawk.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply gossip/minimal/goshawk.go]: Feel free. My point is mainly this: if you change the condition (say, you check other field of `cfg` instead), you might forget to update the log string below accordingly.

File: gossip/minimal/goshawk.go (hunk 1 above)
Comment (pavelkalinnikov) [reply gossip/minimal/goshawk.go]: Not sure I support this, unless this rule is documented somewhere. Standard Go libs don't seem to follow this convention. For example, `http.Server`'s continuous/blocking "run" method is still a verb: https://golang.org/pkg/net/http/#Server.ListenAndServe.

File: gossip/minimal/goshawk.go (hunk 3 above)
Comment (pavelkalinnikov) [reply gossip/minimal/goshawk.go]: Again, feel free. I just got used to see `if err != nil` first thing after any call (precisely for the same "conventions" reason), so seeing something different triggers "why?"s in my head automatically. Also, some linters complain when `err` is ignored.

File: gossip/minimal/goshawk.go (hunk 2 above)
Comment (pavelkalinnikov) [reply gossip/minimal/goshawk.go]: Leaving this for a follow-up PR?

Comment (pavelkalinnikov) [reply gossip/minimal/goshawk.go]: Ah, didn't notice you fixed that. Ignore my comment.

File: gossip/minimal/goshawk.go (hunk 1 above)
Comment (gdbelvin) [reply gossip/minimal/goshawk.go]: I believe Go encourages "*er"  suffixes in interfaces.  ie. `io.Writer` and `io.Reader`.

File: scanner/fetcher.go [hunk 6]
```diff
@@ -230,7 +231,8 @@ func (f *Fetcher) updateSTH(ctx context.Context) error {
 
 		quick := time.Now().Before(quickDeadline)
 		if sth.TreeSize <= lastSize || quick && sth.TreeSize < targetSize {
-			return errors.New("waiting for bigger STH")
+			// Use an explicitly retriable error code.
```
Comment (pavelkalinnikov) [inline scanner/fetcher.go]: `// TODO: Use RetriableError(fmt.Errorf(...)) instead.`

File: gossip/minimal/goshawk.go (hunk 3 above)
Comment (daviddrysdale) [reply gossip/minimal/goshawk.go]: Added `err != nil` arm too.

File: scanner/fetcher.go (hunk 6 above)
Comment (daviddrysdale) [reply scanner/fetcher.go]: Done, assuming that's now the plan for `backoff`

**Pull request merged**